src/arch/riscv/pmp.cc

```diff
@@ -0,0 +1,223 @@
+/*
+ * Copyright (c) 2021 The Regents of the University of California
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met: redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer;
+ * redistributions in binary form must reproduce the above copyright
+ * notice, this list of conditions and the following disclaimer in the
+ * documentation and/or other materials provided with the distribution;
+ * neither the name of the copyright holders nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "arch/riscv/pmp.hh"
+
+#include "arch/generic/tlb.hh"
+#include "arch/riscv/faults.hh"
+#include "arch/riscv/isa.hh"
+#include "base/addr_range.hh"
+#include "base/types.hh"
+#include "cpu/thread_context.hh"
+#include "math.h"
+#include "mem/request.hh"
+#include "params/PMP.hh"
+#include "sim/sim_object.hh"
+
+PMP::PMP(const Params &params) :
+    SimObject(params),
+    maxEntries(params.max_pmp),
+    numRules(0)
+{
+    for (int i=0; i < maxEntries; i++) {
+        PmpEntry entry;
+        pmpTable.emplace_back(entry);
+    }
+}
+
+Fault
+PMP::pmpCheck(const RequestPtr &req, BaseTLB::Mode mode,
+              RiscvISA::PrivilegeMode pmode, ThreadContext *tc)
+{
+    // First determine if pmp table should be consulted
+    if (!shouldCheckPMP(pmode, mode, tc))
+        return NoFault;
+
+    // An access should be successful if there are
+    // no rules defined yet or we are in M mode (based
+    // on specs v1.10)
+    if (numRules == 0 || (pmode == RiscvISA::PrivilegeMode::PRV_M))
+        return NoFault;
+
+    // all pmp entries need to be looked from the lowest to
+    // the highest number
+    int match_index = -1;
```

Comment:
I think this is needed because for range entries you actually need two PMP registers, right? I think it could be worth a comment about this.

Reply:
Assuming that you are referring to TOR mode, the pmpUpdateRule takes care of this and creates a single address range for that particular pmp entry. In pmpCheck match_index is needed just to keep track of which pmp entry the address matched with so that the index can be used to access permission bits from that pmp entry.